pkg/uid/uid.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2015 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// For how the uidshift and uidcount are generated please check:
+// http://cgit.freedesktop.org/systemd/systemd/commit/?id=03cfe0d51499e86b1573d1
+
+package uid
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"time"
+)
+
+const DefaultRangeCount = 0x10000
+
+// A UidRange structure used to set uidshift and its range.
+type UidRange struct {
+	Shift uint32
+	Count uint32
+}
+
+func generateUidShift() uint32 {
+	rand.Seed(time.Now().UnixNano())
+	// we force the MSB to 0 because devpts parses the uid,gid options as int
+	// instead of as uint.
+	// http://lxr.free-electrons.com/source/fs/devpts/inode.c?v=4.1#L189
+	// systemd issue: https://github.com/systemd/systemd/issues/956
+	n := rand.Intn(0x7FFF) + 1
+	uidShift := uint32(n << 16)
+
+	return uidShift
+}
+
+func NewBlankUidRange() *UidRange {
+	return &UidRange{
+		Shift: 0,
+		Count: 0}
+}
+
+func (r *UidRange) SetRandomUidRange(uidCount uint32) {
+	uidShift := generateUidShift()
+	r.Shift = uidShift
+	r.Count = uidCount
+}
+
+func (r *UidRange) ShiftRange(uid uint32, gid uint32) (uint32, uint32, error) {
+	if r.Count > 0 && (uid >= r.Count || gid >= r.Count) {
+		return 0, 0, fmt.Errorf("uid %d or gid %d are out of range %d", uid, gid, r.Count)
+	}
+	if math.MaxUint32-r.Shift < uid || math.MaxUint32-r.Shift < gid {
```

Comment:
Sorry about the confusion with this `0xffffffff`. I somehow thought that `r.Shift` is already an `uint64`, so the result of adding `uint32` to `uint64` would be `uint64`. Which wasn't the case and such addition is a compilation error anyway.